docker/api/client.py

```diff
@@ -381,19 +386,58 @@ def _stream_raw_result(self, response, chunk_size=1, decode=True):
         for out in response.iter_content(chunk_size, decode):
             yield out
 
-    def _read_from_socket(self, response, stream, tty=False):
+    def _read_from_socket(self, response, stream, tty=False, demux=False):
         socket = self._get_raw_response_socket(response)
 
-        gen = None
-        if tty is False:
+        if tty:
+            # When `tty=True`, there's no framing, we just read the raw data
+            # from the socket. It is as if the streams were multiplexed and
+            # spit out on stdout.
+            gen = ((STDOUT, data) for data in socket_raw_iter(socket))
+        else:
             gen = frames_iter(socket)
+
+        if demux:
+            # The generator will output 3-tuples (stdin, stdout, stderr)
+            gen = (self._demux_adaptor(*frame) for frame in gen)
         else:
-            gen = socket_raw_iter(socket)
+            # The generator will output strings
+            gen = (data for (_, data) in gen)
 
         if stream:
             return gen
         else:
-            return six.binary_type().join(gen)
+            # Wait for all the frames, concatenate them, and return the result
+            return self._consume_socket_output(gen, demux=demux)
+
+    @staticmethod
+    def _consume_socket_output(frames, demux=False):
```

Comment:
This method as well as `_demux_adaptor` should probably be in `docker/utils/socket.py`

Reply:
done